Seperate view life cycle method

cmicat opened the pull request:
1. finalize onCreateViewHolder, onBindViewHolder , this fix #3
2. make onCreateHeaderViewHolder, onCreateFooterViewHolder, onBindHeaderViewHolder, onBindFooterViewHolder not abstract, since headers and footers are optional, users may ignore to override those method.
3. seperate onFailedToRecycleView to onFailedToRecycleHeaderView, onFailedToRecycleItemView, onFailedToRecycleFooterView
4. seperate onViewAttachedToWindow to onHeaderViewAttachedToWindow, onItemViewAttachedToWindow, onFooterViewAttachedToWindow
5. seperate onViewDetachedFromWindow to onHeaderViewDetachedFromWindow, onItemViewDetachedFromWindow, onFooterViewDetachedFromWindow
6. seperate onViewRecycled to onHeaderViewRecycled, onItemViewRecycled, onFooterViewRecycled
7. update robolectric dependency to 3.0
8. update mockito dependency 1.10.19
9. update junit dependency to 4.12
10. add powermock 1.6.4 dependency
11. seperate test base class to PowermockTest, RobolectricTestWithMockito, RobolectricTestWithPowermock. Because I can't get spy() work with RobolectricTestWithPowermock, while it's really useful to mock final methods.

I have tested my change in Unit test, in the example project, and in one of my own project.

cmicat committed:
mark onBindViewHolder & onCreateViewHolder as final

cmicat committed:
1. update Robolectric library version to 3.0
2. update JUnit library version to 4.12

cmicat committed:
add PowerMock dependency

cmicat committed:
seperate onFailedToRecycleView

cmicat committed:
seperate onViewDetachedFromWindow

cmicat committed:
seperate onViewRecycled

cmicat committed:
add test code in example app

cmicat committed:
make
onBindFooterViewHolder,
onBindHeaderViewHolder,
onCreateFooterViewHolder,
onCreateHeaderViewHolder
not abstract

cmicat committed:
update readme

cmicat committed:
update style

pedrovgs commented on README.md:
```diff
@@ -22,8 +22,8 @@ public class DragonBallAdapter extends HeaderRecyclerViewAdapter<RecyclerView.Vi
-* 2 - Implement ``onCreateHeaderViewHolder``, ``onCreateItemViewHolder``,``onCreateFooterViewHolder`` , ``onBindHeaderViewHolder``, ``onBindItemViewHOlder`` and ``onBindFooterViewHolder`` to create your ``RecyclerView.ViewHolder`` instances and draw your rows:
-
+* 2 - Implement ``onCreateHeaderViewHolder``, ``onCreateItemViewHolder``,``onCreateFooterViewHolder`` , ``onBindHeaderViewHolder``, ``onBindItemViewHOlder`` and ``onBindFooterViewHolder`` to create your ``RecyclerView.ViewHolder`` instances and draw your rows:<br>
+**if you don't use header or footer, you can ignore corresponding createViewHolder and bindViewHolder method**
```
You have a typo here. This `if` should start with an i but using uppercase.

pedrovgs replied:
Please, do not use `<br>` ad an blank line if needed.

pedrovgs commented on README.md:
````diff
@@ -67,7 +67,63 @@ public class DragonBallAdapter extends HeaderRecyclerViewAdapter<RecyclerView.Vi
 ```
 
-* 3 - Configure your ``RecyclerView`` widget with this layout:
+* 3 - you can use following method to handle view life cycle event:
````
Please, do not change this sentence.

pedrovgs commented on library/build.gradle:
```diff
@@ -1,28 +1,32 @@
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
```
Review you IDE configuration, you've added some whitespace and extra indentation we are not using. You can use this script to configure your IDE https://github.com/square/java-code-styles and then reformat this file.

pedrovgs commented on library/build.gradle:
```diff
@@ -1,28 +1,32 @@
+    testCompile "org.powermock:powermock-module-junit4:1.6.4"
+    testCompile "org.powermock:powermock-module-junit4-rule:1.6.4"
+    testCompile "org.powermock:powermock-api-mockito:1.6.4"
+    testCompile "org.powermock:powermock-classloading-xstream:1.6.4"
```
The same here. We should be careful with the checkstyle configuration.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
  * onCreateViewHolder to create your ViewHolder instances and onBindViewHolder methods to draw your
  * user interface as you wish.
- *
+ * <p>
```
Please, do not add html elements to the javadoc.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
-  private boolean hasItems() {
-    return items.size() > 0;
-  }
+        extends RecyclerView.Adapter<VH> {
```
You've changed the whole file to use 4 spaces instead of 2. Please apply the checkstyle configuration I commented you few comments ago and reformat the code.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+    private F footer;
+    private boolean showFooter = true;
+
+    //<editor-fold desc="seperate onCreateViewHolder">
```
Can we remove this line? We are not using this part of the javadoc api.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+    }
+    //</editor-fold>
+
+    //<editor-fold desc="seperate onBindViewHolder">
```
Remove also this `</editor-fold>  marks` please.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+
+    protected boolean onFailedToRecycleFooterView(VH holder) {
+        return false;
+    }
```
We don't used to use this part of the `RecyclerView` API. If I'm not wrong we can remove these four methods:
- `onFailedToRecycleView`
- `onFailedToRecycleHeaderView`
- `onFailedToRecycleItemView`
- `onFailedToRecycleFooterView`

Remember to update the README.md with these changes.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+
+    protected void onFooterViewRecycled(VH holder) {
+    }
+    //</editor-fold>
```
To keep the library simplicity we can also remove these methods related to: `onViewDetachedFromWindow` and `onViewRecycled`

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+        return hasHeader() && position == 0;
+    }
+
+    public boolean isFooterPosition(int position) {
```
Can you add javadoc to this method?

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -26,183 +27,299 @@
+        return viewType == TYPE_HEADER;
+    }
+
+    protected boolean isFooterType(int viewType) {
```
Can you add javadoc to this method?

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -18,247 +18,695 @@
 public abstract class HeaderRecyclerViewAdapterBaseTest<VH extends RecyclerView.ViewHolder, H, T, F>
-    extends RobolectricTest {
-
-  private static final int TYPE_ITEM = -1;
```
As I said before, apply the java checkstyle and reformat this file again.

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -18,247 +18,695 @@
+        assertEquals(positionZeroItem, adapterWithHeaderAndWithFiveItems.getItem(POSITION_ONE));
+    }
+
+    //<editor-fold desc="onCreateViewHolder test">
```
Remember to remove these comments.

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/testBaseClass/PowermockTest.java:
```diff
@@ -0,0 +1,18 @@
+ * else if(you don't need to feature provided by Powermock, like mock final,private or static method)
+ *     use {@link RobolectricTestWithMockito}
+ * else
+ *     use {@link RobolectricTestWithPowermock}</pre>
```
Please, do not use pseudo code to write the javadoc. Remember to remove the html comments.

pedrovgs commented on library/src/test/java/com/karumi/headerrecyclerview/testBaseClass/PowermockTest.java:
```diff
@@ -0,0 +1,18 @@
+ *
+ */
+@RunWith(PowerMockRunner.class)
+public abstract class PowermockTest {
```
Why are you using powermock?

pedrovgs commented:
Thanks for your PR @cmicat!! I'l continue reviewing your code once you apply the comments I've posted.

pedrovgs commented:
@flipper83 can you also review this PR, please? I need two confirmations before to merge.

cmicat committed:
1. reformat code use square android code style
2. remove onFailedToRecycleView, onViewAttachedToWindow, onViewDetachedFromWindow
3. remove RobolectricTestWithPowermock, because it's buggy.
4. rename RobolectricTestWithMockito back to RobolectricTest

cmicat committed:
update javadoc in DebugHelper.java

cmicat committed:
update DragonBallAdapter.java, according to the library code change

cmicat committed:
update README.md

cmicat committed:
update CONTRIBUTING.md

cmicat committed:
remove html element in HeaderRecyclerViewAdapter's javadoc

cmicat commented:
Since I changed a lot of things, according to advices of @pedrovgs, I rewrite what I did as following.
1. Finalize onCreateViewHolder, onBindViewHolder , this fix #3
2. Make onCreateHeaderViewHolder, onCreateFooterViewHolder, onBindHeaderViewHolder, onBindFooterViewHolder not abstract, since headers and footers are optional, users may ignore to override those method.
3. Seperate onViewRecycled to onHeaderViewRecycled, onItemViewRecycled, onFooterViewRecycled. In onViewRecycled(), I need to free some resources in my item view, or there may be a temporary memory leak, since the resources are bitmaps, the memory leaks is really   dangerous, I can't ignore to recycle it, so I did above changes.
4. Update robolectric dependency to 3.0
5. Update mockito dependency 1.10.19
6. Update junit dependency to 4.12
7. Add powermock 1.6.4 dependency
8. Seperate test base class to PowermockTest, RobolectricTest. Because I need to mock `android.support.v7.widget.RecyclerView.ViewHolder.getAdapterPosition()` in android support library, but it's a final method, Robolectric don't support to mock it. I have tried to use Powermock and Robolectric together, but it's problematic, and always fail for some reason. EG: Both Powermock and Robolectric use a custom classloader, and they will have conflicts.
9. update README.md
10. update CONTRIBUTING.md, add instructions for using SquareAndroid code styles, and Square code inspections in this project.

cmicat commented on library/src/test/java/com/karumi/headerrecyclerview/GridLayoutManagerBuilder.java:
```diff
@@ -2,11 +2,12 @@
 public class GridLayoutManagerBuilder {
 
-  Context context = Robolectric.application;
+  Context context = RuntimeEnvironment.application;
```
It's needed because I update Robolectric to 3.0.

cmicat commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -261,4 +353,6 @@ public void shouldReturnFirstItemAsItemAtPositionOneIfHeaderWasConfiguredAndTher
 
   protected abstract List<T> givenAListWithFiveItems();
+
+  protected abstract VH givenAMockViewHolder();
```
`givenAViewHolder()` return a not mockable object, I am not able use Powermock to mock `ViewHolder.getAdapterPosition()`

cmicat commented on library/src/test/java/com/karumi/headerrecyclerview/testBaseClass/RobolectricTest.java:
```diff
@@ -14,16 +14,17 @@
  */
-@Config(emulateSdk = 18) @RunWith(RobolectricTestRunner.class)
-public abstract class RobolectricTest {
+@Config(constants = BuildConfig.class,
```
update Robolectric to 3.0

cmicat commented on sample/src/main/java/com/karumi/headerrecyclerview/sample/DragonBallAdapter.java:
```diff
@@ -72,4 +74,16 @@
+
+  @Override protected void onFooterViewRecycled(RecyclerView.ViewHolder holder) {
+    DebugHelper.logMethod(LOG_TAG);
+  }
```
use `DebugHelper.logMethod(LOG_TAG);` to log current method.

cmicat commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -119,45 +187,58 @@ public T getItem(int position) {
-  protected abstract void onBindItemViewHolder(VH holder, int position);
-
-  protected abstract void onBindFooterViewHolder(VH holder, int position);
-
```
Just moved below `onCreateViewHolder(ViewGroup parent, int viewType)` and `onBindViewHolder(VH holder, int position)`.

cmicat commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -59,10 +59,26 @@
+    return null;
+  }
+
+  /**
```
Header and footer are optional, so don't use `abstract` modifier to force user to override `onCreateHeaderViewHolder` and `onCreateFooterViewHolder`

cmicat commented on library/src/test/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapterBaseTest.java:
```diff
@@ -18,22 +18,27 @@
+@PrepareForTest(RecyclerView.ViewHolder.class)
 public abstract class HeaderRecyclerViewAdapterBaseTest<VH extends RecyclerView.ViewHolder, H, T, F>
-    extends RobolectricTest {
+    extends PowermockTest {
```
This test class don't need fake Android runtime, but to mock some final method, so I extends PowermockTest .

flipper83 replied:
we don't use powermock, because people can write ugly test using it. We prefer use or RobolectricTest if the class that we can test is from the framework, or new android test runner that all methods are not final and you can mock it.

cmicat replied:
I tried use `android.support.test.runner.AndroidJUnit4` test runner and mockito, run in "Android Instrumentation Test" to mock `android.support.v7.widget.RecyclerView.ViewHolder.getAdapterPosition()`, but not work. It's a final method, and it seems that google do not provide a mockable version of support-v7 library. I look at  [Robolectric Add-On Modules page](http://robolectric.org/using-add-on-modules/), but they do not provide support-v7 mockable version, too. 
I'm not able to refactor the system to introduce a wrapper class that I control, because holder parameter of `onViewRecycled(VH holder)` is passed in by RecyclerView api, and it must be a   `RecyclerView.ViewHolder`. 

So I only have two choices:
1. don't unit test this method.
2. use powermock to mock final method.

And I failed to integrate Powermock and Robolectric together, as I mentioned before." Both Powermock and Robolectric use a custom classloader, and they will have conflicts."
Another error is Powermock will never be able to mock final method when use with Robolectric, it just delegate Powermock.mock() call to Mockito.mock(), and mockito gives me an Exception when I attempt mocking final methods.

pedrovgs replied:
I think that the usage of powermock is correct here.

flipper83 replied:
ok, If we don't have other way to fix this one, go ahead.

cmicat replied:
Ok, I succeed integrate Powermock and Robolectric together, I should make a unified RobolectricTest which supports to use powermock.

But Powermock and Robolectric integrated together isn't really stable, they often throw a lot of Exceptions, I need to test more. After that, I will push my changes.

pedrovgs commented on sample/src/main/java/com/karumi/headerrecyclerview/sample/util/DebugHelper.java:
```diff
@@ -0,0 +1,27 @@
+ * Debug utility class
+ */
+public class DebugHelper {
+  private static String getCallMethodAndLine() {
```
We don't need this `DebugHelper` class. Just add a simple log on `DragonBallAdapter` methods :)

cmicat replied:
Ok, it's an overkill for an example app. And may confuse users.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -46,7 +46,7 @@
    * based on the view type param.
    */
-  @Override public VH onCreateViewHolder(ViewGroup parent, int viewType) {
+  @Override public final VH onCreateViewHolder(ViewGroup parent, int viewType) {
```
This method should not be declared as final. Doing this we are restricting the component usage based on inheritance.

cmicat replied:
You said:
"These methods just check if the view type or the position is related to a header. Making these method finals the library clients can override onCreateItemViewHolder and onBindItemViewHolder to implement custom bindings."
here https://github.com/Karumi/HeaderRecyclerView/issues/3#issuecomment-106264010

So did I misunderstand your meaning?

If the user want to use different view types but the same data structure, they can simply dectect what item type it is in onCreateItemViewHolder, and create the needed view. But if they want to create something like header2 or footer2, mark onCreateViewHolder as final will compel them to put header2 and footer2 to items, even though they don't have any common point with items. 

So declare onCreateViewHolder as final will restrict users to add additional special view types(not items), right?

Em, therefore, #3 should be closed, and I will also make onCreateViewHolder not final.
(∩_∩)

pedrovgs replied:
You are completely right, my fault.

cmicat replied:
So users don't need to create header2 or footer2?

pedrovgs replied:
If they need it they should not use this library. This library offers one header and one footer, the rest of the features should be implemented using another more complex library like Renderers.

pedrovgs commented on library/src/main/java/com/karumi/headerrecyclerview/HeaderRecyclerViewAdapter.java:
```diff
@@ -59,10 +59,26 @@
    * on the position param.
    */
-  @Override public void onBindViewHolder(VH holder, int position) {
+  @Override public final void onBindViewHolder(VH holder, int position) {
```
Neither this one. Declaring this method as final, as I said before, we are restricting the component usage based on inheritance. Removing this final modifier we can also remove the usage of powermock :)

pedrovgs replied:
Forget this comment, sorry.

pedrovgs commented:
Just one comment pending to be reviewed!!! Replace the DebugHelper class with a simple log and we are done!!

cmicat committed:
Integrate Powermock and Robolectric together.

cmicat committed:
remove DebugHelper

cmicat committed:
update README.md "Libraries used in this project" section

cmicat commented:
Ok, done. Thank you for your excellent work with this library.

pedrovgs commented:
Thanks @cmicat you have my :+1:

pedrovgs commented:
@flipper83 do you want to add something else?

flipper83 commented:
yo have my :+1: too
